### PR TITLE
Migrate `incr-add-rust-src-component` to rmake

### PR DESCRIPTION
**Pull request by jieyouxu**
This PR partially supersedes #128562, and ports the Makefile-based `tests/run-make/incr-add-rust-src-component` to use rmake.rs infra.

Part of #121876.

This run-make test is a regression test for https://github.com/rust-lang/rust/issues/70924. It (tries to) checks that if we add the `rust-src` component in between two incremental compiles, that the compiler doesn't ICE on the second invocation.

- Original issue:https://github.com/rust-lang/rust/issues/70924
- Fix PR: https://github.com/rust-lang/rust/pull/72767
- PR adding this regression test: https://github.com/rust-lang/rust/pull/72952

However, the Makefile version of this used `$SYSROOT/lib/rustlib/src/rust/src/libstd/lib.rs`, but that actually got moved around and reorganized over the years. As of Dec 2024, the `rust-src` component is more like (specific for our purposes):

```
$SYSROOT/lib/rustlib/src/rust/
    library/std/src/lib.rs
    src/
```

However, this run-make test is ancient and it exercises incr-comp system logic. I'm not sure if this test would actually catch the original regression.

This PR was co-authored with @Oneirical.

r? incremental

try-job: i686-msvc
try-job: x86_64-mingw-1
try-job: x86_64-msvc
try-job: aarch64-apple

### CODE REVIEW

**Comment by jieyouxu** [comment]
Ah. The permission failure on msvc is genuine, blocked on #126334.

@rustbot blocked

File: tests/run-make/incr-add-rust-src-component/rmake.rs [hunk 1]
```diff
@@ -0,0 +1,131 @@
+//! Regression test for rust-lang/rust#70924. Check that if we add the `rust-src` component in
+//! between two incremental compiles, that the compiler doesn't ICE on the second invocation.
+//!
+//! This test uses symbolic links to save testing time.
+//!
+//! The way this test works is that, for every prefix in `root/lib/rustlib/src`, link all of prefix
+//! parent content, then remove the prefix, then loop on the next prefix. This way, we basically
+//! create a copy of the context around `root/lib/rustlib/src`, and can freely add/remove the src
+//! component itself.
```
Comment (jieyouxu) [inline tests/run-make/incr-add-rust-src-component/rmake.rs:9]: Question: ... is this test actually testing what it thinks it's testing?

File: tests/run-make/incr-add-rust-src-component/Makefile [hunk 2]
```diff
@@ -1,45 +0,0 @@
-# ignore-cross-compile
-include ../tools.mk
-
-# rust-lang/rust#70924: Test that if we add rust-src component in between two
-# incremental compiles, the compiler does not ICE on the second.
-
-# This test uses `ln -s` rather than copying to save testing time, but its
-# usage doesn't work on windows. So ignore windows.
-
-# ignore-windows
```
Comment (jieyouxu) [inline tests/run-make/incr-add-rust-src-component/Makefile:10]: Remark: by usage of pure rust test file, we're able to expand the test coverage of this to Windows as well. The symlink handling was fixed and improved in https://github.com/rust-lang/rust/pull/134659.

File: tests/run-make/incr-add-rust-src-component/rmake.rs (hunk 1 above)
Comment (wesleywiser) [reply tests/run-make/incr-add-rust-src-component/rmake.rs:9]: I _think_ so... in either case, your translation appears to preserve the intended behavior of the test (but perhaps that behavior itself isn't entirely correct?)

**Review by wesleywiser — APPROVED** [review]

**Pull request merged**